infrastructure/src/main/java/org/corfudb/infrastructure/management/CompleteGraphAdvisor.java

```diff
@@ -0,0 +1,134 @@
+     */
+    @Override
+    public Optional<NodeRank> failedServer(
+            ClusterState clusterState, List<String> unresponsiveServers, String localEndpoint) {
+
+        log.trace("Detecting failed nodes for: ClusterState= {} unresponsive servers= {}",
+                clusterState, unresponsiveServers
+        );
+
+        ClusterGraph graph = ClusterGraph.transform(clusterState);
+
+        ClusterGraph symmetric = graph.toSymmetric();
+        Optional<NodeRank> maybeDecisionMaker = symmetric.getDecisionMaker();
+
+        if (!maybeDecisionMaker.isPresent()) {
+            return Optional.empty();
+        }
+
+        NodeRank decisionMaker = maybeDecisionMaker.get();
+        if (!decisionMaker.is(localEndpoint)) {
+            String message = "The node can't be a decision maker, skip operation. Decision maker node is: {}";
+            log.debug(message, decisionMaker);
+            return Optional.empty();
+        }
+
+        Optional<NodeRank> maybeFailedNode = symmetric.findFailedNode();
+        if (!maybeFailedNode.isPresent()) {
+            return Optional.empty();
+        }
+
+        NodeRank failedNode = maybeFailedNode.get();
+
+        if (decisionMaker.equals(failedNode)) {
+            log.error("Decision maker and failed node are same node: {}", decisionMaker);
+            return Optional.empty();
+        }
+
+        if (unresponsiveServers.contains(failedNode.getEndpoint())) {
+            return Optional.empty();
+        }
+
+        log.debug("Failed node found: {}", failedNode);
+        return Optional.of(failedNode);
+    }
+
+    /**
+     * Provide a server considered to have healed in the Corfu cluster according to
+     * the COMPLETE_GRAPH implementation of algorithm for
+     * {@link ClusterType}.
+     * <p>
+     * The node can heal only itself. The node responsible only for itself, can't heal other nodes.
+     * It simplifies healing algorithm and guaranties that if node became available it mark itself as a responsible
+     * node in the layout. It helps us to simplify analysis/debugging process and brings simple and reliable algorithm
+     * for healing process.
+     *
+     * @param clusterState        represents the state of connectivity amongst the Corfu cluster
+     *                            nodes from a node's perspective.
+     * @param unresponsiveServers unresponsive servers in a layout.
+     * @return a {@link List} of servers considered as healed according to the underlying
+     * {@link ClusterType}.
+     */
+    @Override
+    public Optional<NodeRank> healedServer(
+            ClusterState clusterState, List<String> unresponsiveServers, String localEndpoint) {
+
+        log.trace("Detecting the healed nodes for: ClusterState: {} unresponsive servers: {}",
+                clusterState, unresponsiveServers
+        );
+
+        if (unresponsiveServers.isEmpty()) {
+            log.trace("All nodes responsive. Nothing to heal");
```

Comment:
I try to avoid `if` blocks with multiple conditions and enclosed if statements. The idea is - use only simple if blocks with a return statement. 
When you read the method, you see all "unsuccessful" cases, and domain logic in the method can be executed only if all unsuccessful cases passed. Domain logic tends to stay at the of the method, it's not in an enclosed method, you always know that most important logic at a particular place of the method.

Reply:
Ok, got it, sounds reasonable. For this particular case, I leave it as is because of different log levels for those conditions. In general, I will reduce the number of `if` blocks if there is no difference between them.